File: app/tabs.py
from __future__ import annotations

from typing import Any
from urllib.parse import urlparse

MAX_TABS = 80
MAX_TITLE = 300
MAX_URL = 2000
ALLOWED_SCHEMES = frozenset({"http", "https"})


def sanitize_tabs(raw: list[Any] | None) -> list[dict[str, Any]]:
    """Keep only currently-open http(s) titles+URLs. Never persist file paths."""
    cleaned: list[dict[str, Any]] = []
    seen: set[str] = set()
    for item in raw or []:
        if not isinstance(item, dict):
            continue
        url = str(item.get("url") or "").strip()[:MAX_URL]
        title = str(item.get("title") or "").strip()[:MAX_TITLE]
        if not url:
            continue
        parsed = urlparse(url)
        if parsed.scheme not in ALLOWED_SCHEMES:
            continue
        if not parsed.netloc:
            continue
        key = url
        if key in seen:
            continue
        seen.add(key)
        window_id = item.get("window_id", item.get("windowId"))
        cleaned.append(
            {
                "title": title or url,
                "url": url,
                "active": bool(item.get("active")),
                "pinned": bool(item.get("pinned")),
                "window_id": window_id if isinstance(window_id, int) else None,
            }
        )
        if len(cleaned) >= MAX_TABS:
            break
    return cleaned

File: src-tauri/src/tabs.rs
use serde::{Deserialize, Serialize};
use url::Url;

pub const ALLOWED_SCHEMES: [&str; 2] = ["http", "https"];
pub const MAX_TABS: usize = 80;
pub const MAX_TITLE: usize = 200;
pub const MAX_URL: usize = 2048;

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Tab {
    #[serde(default)]
    pub title: String,
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub active: bool,
    #[serde(default)]
    pub pinned: bool,
    #[serde(default)]
    pub window_id: Option<i64>,
}

pub fn sanitize_tabs(raw: &[Tab]) -> Vec<Tab> {
    let mut out = Vec::new();
    let mut seen = std::collections::HashSet::new();
    for tab in raw {
        if out.len() >= MAX_TABS {
            break;
        }
        let url = tab.url.trim();
        if url.len() > MAX_URL {
            continue;
        }
        let Ok(parsed) = Url::parse(url) else {
            continue;
        };
        if !ALLOWED_SCHEMES.contains(&parsed.scheme()) {
            continue;
        }
        if parsed.host_str().unwrap_or("").is_empty() {
            continue;
        }
        if !seen.insert(url.to_string()) {
            continue;
        }
        let mut title = tab.title.trim().to_string();
        title.truncate(MAX_TITLE);
        out.push(Tab {
            title,
            url: url.to_string(),
            active: tab.active,
            pinned: tab.pinned,
            window_id: tab.window_id,
        });
    }
    out
}
